contracts/SpokeV3PoolPeriphery.sol

```diff
@@ -87,40 +90,47 @@ contract SpokePoolV3Periphery is Lockable, MultiCaller {
     error InvalidSpokePool();
     error InvalidSwapToken();
+    error InvalidExchange();
 
     /**
      * @notice Construct a new SwapAndBridgeBase contract.
-     * @param _allowedSelectors Function selectors that are allowed to be called on the exchange.
+     * @dev Is empty and all of the state variables are initialized in the initialize function
+     * to allow for deployment at a deterministic address via create2, which requires that the bytecode
+     * across different networks is the same. Constructor parameters affect the bytecode so we can only
+     * add parameters here that are consistent across networks.
      */
-    constructor(bytes4[] memory _allowedSelectors) {
-        for (uint256 i = 0; i < _allowedSelectors.length; i++) {
-            allowedSelectors[_allowedSelectors[i]] = true;
-        }
-    }
+    constructor() {}
 
     /**
      * @notice Initializes the SwapAndBridgeBase contract.
      * @param _spokePool Address of the SpokePool contract that we'll submit deposits to.
      * @param _wrappedNativeToken Address of the wrapped native token for the network this contract is deployed to.
-     * @param _exchange Address of the exchange where tokens will be swapped.
+     * @param exchanges Array of exchange addresses and their allowed function selectors.
      * @dev These values are initialized in a function and not in the constructor so that the creation code of this contract
-     * is the same across networks with different addresses for the wrapped native token, the exchange this contract uses to
-     * swap and bridge, and this network's corresponding spoke pool contract. This is to allow this contract to be deterministically
-     * deployed with CREATE2.
-     * @dev This function can be front-run by anybody, so it is critical to check that the `spokePool`, `wrappedNativeToken`, and `exchange`
-     * values used in the single call to this function were passed in correctly before enabling the usage of this contract.
+     * is the same across networks with different addresses for the wrapped native token and this network's
+     * corresponding spoke pool contract. This is to allow this contract to be deterministically deployed with CREATE2.
+     * @dev This function can be front-run by anybody, so it is critical to check that the values used in the
+     * single call to this function were passed in correctly before enabling the usage of this contract.
      */
     function initialize(
         V3SpokePoolInterface _spokePool,
         WETH9Interface _wrappedNativeToken,
-        address _exchange
+        WhitelistedExchanges[] calldata exchanges
     ) external {
         if (initialized) revert ContractInitialized();
         initialized = true;
 
+        if (!address(_spokePool).isContract()) revert InvalidSpokePool();
         spokePool = _spokePool;
         wrappedNativeToken = _wrappedNativeToken;
-        exchange = _exchange;
+        for (uint256 i = 0; i < exchanges.length; i++) {
```

Comment:
Very nitty, but I think we can optimize gas here if we do something like:
```
WhitelistedExchanges calldata _exchange;
uint256 nExchanges = exchanges.length;
for (uint256 i = 0; i < nExchanges; i++) {
  _exchange = exchanges[i];
  ...
}
```
Although, this _is_ just an init function, so probably not necessary.

Reply:
fair points! https://github.com/across-protocol/contracts/pull/777/commits/c1b6d4f39b626f1b1371a7b24854156cbb33be83

Reply:
Do you also think it' worthwhile to declare the exchanges as `calldata` over `memory`. Right now, it doesn't look like it needs to be memory, and I think that might save gas if we want to be super particular.